Allow changing the reader/writer threads count at runtime

eifrah-aws opened the pull request:
With this PR, user can control the threads counts
using

```
config set search.reader-threads <value>
config get search.reader-threads

config set search.writer-threads <value>
config get search.writer-threads
```

In addition, refactored the configuration registration mechanism. Calling `Builder<type>("my-config",...).Build()` will create an instance of configuration entry named `my-config` that can be controlled from the Valkey's `CONFIG SET|GET search.my-config` command AND from the command line by passing `--my-config ...` 

**Example 1:**

In order to register a configuration entry with Valkey, named: `search.hnsw-block-size`, the developer needs to place similar lines of code in one of his source files:

```c++
// Include the vmsdk::config namespace
#include "vmsdk/src/module_config.h"

...

// Controls the HNSW resize increments
namespace config = vmsdk::config;

static auto hnsw_block_size =
    config::NumberBuilder("hnsw-block-size",      // name
                          kHNSWDefaultBlockSize,  // default size
                          kHNSWMinimumBlockSize,  // min size
                          UINT_MAX)               // max size
        .WithValidationCallback(ValidateHNSWBlockSize)
        .Build();

// Define some boolean flag and initialize it with false
// we can later change it using `config set search.my-bool-flag yes`
auto my_bool_flag =
    config::BooleanBuilder("my-bool-flag", false).Build();
```

With the above example, it is now also possible to pass these configuration from the
command line like this:

```bash
valkey-server "--loadmodule libsearch.so --my-bool-flag yes --hnsw-block-size 20480"
```

**Example 2:**

We can also register a configuration entry with an action to take when the configuration item is modified:

```c++
// Include the vmsdk::config namespace
#include "vmsdk/src/module_config.h"

...

namespace config = vmsdk::config;

// Controls the number of reader threads
static auto reader_threads_count =
    config::NumberBuilder(
        kReaderThreadsConfig,  // name
        kDefaultThreadsCount,  // default size
        1,                     // min size
        kMaxThreadsCount)      // max size
        .WithModifyCallback(   // set an "On-Modify" callback
            [](long long new_value) {
              UpdateThreadPoolCount(
                  ValkeySearch::Instance().GetReaderThreadPool(), new_value);
            })
        .Build();
```

Note that the callback is called if you set it programmatically like this:

```c++
VMSDK_RETURN_IF_ERROR(reader_threads_count->SetValue(100));
```

or using `valkey-cli`:

```bash
config set search.reader-threads 100
```


![threads](https://github.com/user-attachments/assets/b47ad66f-a5ee-49ed-ac0f-9ff32397ce30)

allenss-amazon commented on vmsdk/src/module_config.cc:
```diff
@@ -0,0 +1,150 @@
+                               void *priv_data, RedisModuleString **err) {
+  auto entry = reinterpret_cast<Number *>(priv_data);
+  CHECK(entry) << "null private data for configuration Number entry.";
+  entry->SetValue(value);
```
This code fails to allow some kind of "validation" function. IMO, an optional validation function should be provided.

eifrah-aws replied:
Done. The validation is now done internally

allenss-amazon reviewed (commented):
1. Need to provide some form of optional validation function. Just having ranges is insufficient for the general case.
2. Should provide some mechanism for adding in the configuration flags like REDISMODULE_CONFIG_IMMUTABLE, etc. 
3. Need some kind of full end to end test, i.e., python based test

allenss-amazon commented on src/valkey_search.cc:
```diff
@@ -65,17 +64,87 @@
+  if (new_value < kHNSWDefaultBlockSize || new_value > UINT32_MAX) {
+    if (err) {
+      *err = RedisModule_CreateStringPrintf(
+          nullptr, "Block size must be between %u and %u",
```
Isn't this dead code now?

eifrah-aws replied:
Still using it in the part we I we set the block size from the command line argument.

allenss-amazon commented on src/valkey_search.cc:
```diff
@@ -65,17 +64,87 @@
+namespace config = vmsdk::config;
+
+// Controls the HNSW resize increments
+static auto hnsw_block_size = config::Builder<config::Number, long long>(
```
Isn't the first type redundant with the second type?

eifrah-aws replied:
Agreed. We can remove it since in our case, there are only limited number of item types (long long, bool, string & flags).

allenss-amazon commented on vmsdk/src/module_config.h:
```diff
@@ -0,0 +1,246 @@
+  FRIEND_TEST(Builder, ConfigBuilder);
+};
+
+template <typename T, typename ValkeyT>
+class ConfigBuilder {
```
I think ConfigBuilder should only have a single type, here: "ValkeyT", the other type "T" is a funciton of ValkeyT and can be computed accordingly.

eifrah-aws replied:
Done

allenss-amazon commented on src/valkey_search.cc:
```diff
@@ -493,16 +573,12 @@ absl::Status ValkeySearch::LoadOptions(RedisModuleCtx *ctx,
+          absl::StrFormat("Could not set %s. Invalid value: %u",
+                          kHNSWBlockSizeConfig, options.hnsw_block_size));
     }
+    hnsw_block_size->SetValue(options.hnsw_block_size);
```
In my CR the command line parsing was also handled by the config variables. I thought that was included here but apparently not. I think it should be.

eifrah-aws replied:
I was not aiming of making it part of this PR. But I can add it.

eifrah-aws replied:
Done

yairgott commented on vmsdk/src/module_config.h:
```diff
@@ -0,0 +1,246 @@
+    current_value_.store(val, std::memory_order_relaxed);
+  }
+
+  bool default_value_ = false;
```
nit: here and other places, prefer using: 
`bool default_value_{false}`

yairgott commented on vmsdk/src/module_config.h:
```diff
@@ -0,0 +1,246 @@
+template <typename T>
+class ConfigBase : public Registerable {
+ public:
+  using OnModifyCB = std::function<void(T)>;
```
Prefer absl::AnyInvocable over std::function

yairgott replied:
It's the same concept as `std::move_only_function`. It ensures that the function will always be moved instead of copied. This has value when the function stores callables  copyable and it's good habit to use it in general.

eifrah-aws commented on src/valkey_search.cc:
```diff
@@ -65,17 +64,87 @@
+/// Check that the new value for configuration item `hnsw-block-size` confirms
+/// to the allowed values.
+bool ValidateHNSWBlockSize(long long new_value, RedisModuleString **err) {
+  if (new_value < kHNSWDefaultBlockSize || new_value > UINT32_MAX) {
```
1. I am removing this method as it is not needed anymore
2. The callback will not get called to begin with, it will be blocked by valkey (since we set a minimal value of `kHNSWDefaultBlockSize`

eifrah-aws replied:
For now, I set to to `0`. We can change this later if needed

murphyjacob4 commented on src/valkey_search_options.cc:
```diff
@@ -0,0 +1,222 @@
+          .Build();
+}
+
+/// If user passed "--threads" it triumphs any value provided by
```
For --threads: would it make sense to remove this for the 1.0 release? I think it is only needed for backwards compatibility with the original Google deployment.

Same goes for the --use-coordinator no argument special case: should we make this conform before 1.0 while we have the chance?

eifrah-aws replied:
About the `--threads` I can remove it, but I would prefer doing it in a separate PR (there are multiple unit tests + integration tests that are using it). There are also some tests such as `threads_win` that check that "--threads" overcome "--reader/writer-threads".

`--use-coordinator` is indeed a different beast, as it is being used as a boolean flag without Valkey convention. Fixing it will remove some code that I added for it

allenss-amazon replied:
I had originally assumed that 1.0 would ship before this work reached completion. With the delays in 1.0, my desire to put this into 1.0 has climbed. I feel like we're close enough to being able to merge all of this and avoid the legacy issues. So I'd vote to delay 1.0 to complete this -- under the assumption that we're only talking about a day or few.

murphyjacob4 replied:
Yeah - definitely shouldn't be lumped with this PR, since we wouldn't want to block 1.0 on this PR that is more of a feature. But if we want to do breaking changes - now is the time

murphyjacob4 replied:
@yairgott wdyt?

murphyjacob4 replied:
Not opposed to blocking 1.0 on this - and cleaning up the configs as part of this

yairgott replied:
Agreed. Lets cleanup the config and include this PR in 1.0.

yairgott replied:
On second thought, I think it makes more sense to create a dedicated PR focused solely on cleaning up the config, rather than rushing a larger change and potentially risking the stability of 1.0.
Thoughts?

eifrah-aws replied:
95% of this PR is focusing on improving the configuration mechanism, so in theory it should have no effect on the stability of the module once the module is up and running. The only thing I can think of is the ability to change the number of threads during runtime (which we can prevent by adding the `IMMUTABLE` flag to these configuration parameters)

As a side note, the level of confidence I have atm with the module comes from the integration tests + unit testing + playing with it locally. And this is true for both cases: with or without this PR.

I vote to include it.

yairgott commented on src/valkey_search_options.cc:
```diff
@@ -0,0 +1,222 @@
+/// Check that the new value for configuration item `hnsw-block-size` confirms
+/// to the allowed values.
+absl::Status ValidateHNSWBlockSize(long long new_value) {
+  if (new_value < kHNSWMinimumBlockSize || new_value > UINT32_MAX) {
```
why new_value should be lower than kHNSWMinimumBlockSize ?

Though, new_value should be greater than 0.

eifrah-aws replied:
1. `kHNSWMinimumBlockSize` is `0` (line 40 on this file)
2. Valkey will not pass the value to the module if it is smaller than the minimal value defined when we first registered the configuration. So this validation basically mimics that behavior if a developer adds a call somewhere in the code to `hnsw_block_size->SetValue(new_value)` or if it passes it from the command line `--loadmodule ... --hnsw-block-size -10`.

yairgott commented on src/valkey_search.cc:
```diff
@@ -143,7 +91,11 @@ void ValkeySearch::InitInstance(std::unique_ptr<ValkeySearch> instance) {
+}
+
+void ValkeySearch::SetHNSWBlockSize(uint32_t block_size) {
+  Options::Instance().GetHNSWBlockSize().SetValue(block_size).IgnoreError();
```
Here and other places, I don't see a reason why we would want to ignore errors, at minimum we should log those.

eifrah-aws replied:
Fixed

murphyjacob4 commented on src/valkey_search_options.cc:
```diff
@@ -0,0 +1,222 @@
+constexpr absl::string_view kHNSWBlockSizeConfig{"hnsw-block-size"};
+constexpr absl::string_view kReaderThreadsConfig{"reader-threads"};
+constexpr absl::string_view kWriterThreadsConfig{"writer-threads"};
+constexpr absl::string_view kUseCoordinator{"use-coordinator"};
```
I was talking with Yair offline - do you think we should just remove --use-coordinator in favor of just enabling/disabling it based on cluster mode?

allenss-amazon replied:
I'm struggling to come up with ANY scenario where having --use-coordinator is something that we want to support in user deployments. Does it have any use in testing? I don't think so, I vote to remove it.

yairgott replied:
I'm also supportive removing it at this point otherwise it would be harder to do so in the future. The single slot index suggested by Allen seems to address the scenario mentioned above.

eifrah-aws replied:
Lets do this on a separate PR then:
- Remove `--threads`
- Remote `--use-coordinator` and rely on the cluster mode (at least until @allenss-amazon single slot index proposal will land)

Thoughts? @yairgott @murphyjacob4

yairgott commented on vmsdk/src/module_config.cc:
```diff
@@ -0,0 +1,300 @@
+      if (!iter.HasNext()) {
+        return absl::NotFoundError(absl::StrFormat(
+            "Command line argument `%s` is missing a value", sv_key));
+      } else {
```
nit: save the need to indent by skipping the `else` when the `if` always returns.

yairgott commented on vmsdk/src/module_config.h:
```diff
@@ -0,0 +1,349 @@
+  }
+
+  std::unique_ptr<ConfigBase<ValkeyT>> Build() {
+    return std::unique_ptr<ConfigBase<ValkeyT>>{config_};
```
I see a few issues with the current approach:
1. `Build()` can be called multiple times, and on the second call it might return a `unique_ptr` to a dangling pointer.
2. It assumes that `Build()` will always be called once; otherwise, *config_ is leaked.
3. `ModuleConfigManager` assumes that its stored config items are valid and not deleted.

I wonder if we can address all of this by storing the config as a `shared_ptr` both internally and within `ModuleConfigManager`, and have `Build()` return a `shared_ptr` as well.

yairgott commented on vmsdk/src/module_config.h:
```diff
@@ -0,0 +1,349 @@
+    return absl::OkStatus();
+  }
+
+  T GetValue() const { return GetValueImpl(); }
```
I think that it would make better sense to return `const T&` which could be assigned to a T by the caller.

eifrah-aws replied:
I don't think its always possible to do so. For example, the concrete class `Number` uses this:

```c++
  long long GetValueImpl() const override {
    return current_value_.load(std::memory_order_relaxed);
  }
```

We can't really return a `const long long&` here.

allenss-amazon commented on vmsdk/src/module_config.h:
```diff
@@ -0,0 +1,371 @@
+    ModuleConfigManager::Instance().RegisterConfig(this);
+  }
+
+  ~ConfigBase() override = default;
```
shoud this unregister itself?

eifrah-aws replied:
Not really needed (the class is removed from the queue by the `ConfigModuleManager` when it does the actual registration. Anyways, added.

allenss-amazon commented on src/valkey_search_options.cc:
```diff
@@ -0,0 +1,232 @@
+    static_cast<int>(LogLevel::kWarning), static_cast<int>(LogLevel::kNotice),
+    static_cast<int>(LogLevel::kVerbose), static_cast<int>(LogLevel::kDebug)};
+
+void Options::DoInitialize() {
```
These are singletons. What is the point of putting them in an object that's manually created/destroyed. It's a bunch of extra verbiage with no point. Why can't these just be global variables? Much cleaner.

allenss-amazon replied:
Making the next 100 configurables harder to use doesn't justify making the testing easier to use now.

eifrah-aws commented on vmsdk/src/module_config.h:
```diff
@@ -0,0 +1,371 @@
+  using OnModifyCB = std::function<void(T)>;
+  using ValidateCB = std::function<absl::Status(const T)>;
+
+  ConfigBase(std::string_view name) : Registerable(name) {
```
Done

allenss-amazon reviewed (changes requested):
The redesign of the vmsdk configuration subsystem to require that all configs be dynamically created during module OnLoad initialization is exactly the opposite direction of where I wanted to go with this. 

The original concept was that configurables would be declared at file level and be initialized automatically. We want to encourage developers to create configurables for any variable that has a particular value, i.e., favor configurables over #define or const variables.

The key issue that I feel has been lost is the ease-of-use. You should be able to add a configurable in one line of code. With this design, it's approaching 10 lines of code. Also, it encourages bad modularization as the temptation is for developers to add another element to the Options structure, which is horribly wordy and will become yet another global dependency rathole.

eifrah-aws commented on vmsdk/src/module_config.h:
```diff
@@ -0,0 +1,379 @@
+/// // Create a Valkey configuration item named "config-name" of type `long
+/// // long`
+/// static auto my_config =
+///    config::Builder<long long>(
```
Fixed

eifrah-aws commented:
> The redesign of the vmsdk configuration subsystem to require that all configs be dynamically created during module OnLoad initialization is exactly the opposite direction of where I wanted to go with this.
> 
> The original concept was that configurables would be declared at file level and be initialized automatically. We want to encourage developers to create configurables for any variable that has a particular value, i.e., favor configurables over #define or const variables.
> 
> The key issue that I feel has been lost is the ease-of-use. You should be able to add a configurable in one line of code. With this design, it's approaching 10 lines of code. Also, it encourages bad modularization as the temptation is for developers to add another element to the Options structure, which is horribly wordy and will become yet another global dependency rathole.

I agree with the overall approach (ease-of-use), but I don't agree with the "static globals" approach. Anyways, and in-order to progress, I will "disgree-and-commit" here ;) so:

I implemented your suggestion and removed the manual instantiation of the `ModuleConfig` & removed `Options` (it is now a pure namespace `options`) 

This allows static variables to be registered before `OnLoad`.
